srk/src/pmfa.ml

```diff
@@ -0,0 +1,493 @@
+      Memo.memo (fun (arr, read) -> 
+          Hashtbl.add func_consist_reqs arr read;
+          let sym = mk_symbol srk `TyInt in
+          nuqr_syms := Symbol.Set.add sym !nuqr_syms;
+          mk_const srk sym)
+    in
+    (* TODO: Make sure that array reads normalized for efficiency; don't want
+     * seperate symbol for a[x + y] vs a[y + x]*)
+    let rec termalg = function
+      |  `Select (a, i) -> 
+        if ArithTerm.equal i uq_term 
+        then (mk_const srk (uq_read a))
+        else (non_uq_read (a, i) :> ('a, typ_arith) expr)
+      | `Ite (cond, bthen, belse) ->
+        mk_ite srk (Formula.eval srk formalg cond) bthen belse
+      | open_term -> ArithTerm.construct srk open_term 
+    and formalg = function
+      | `Atom (`Arith (`Eq, x, y)) -> 
+        let lhs = (ArithTerm.eval srk termalg x) in
+        let rhs = (ArithTerm.eval srk termalg y) in
+        mk_eq srk  lhs rhs   
+      | `Atom (`Arith (`Leq, x, y)) ->
+        mk_leq srk (ArithTerm.eval srk termalg x) (ArithTerm.eval srk termalg y)
+      | `Atom (`Arith(`Lt, x, y)) -> 
+        mk_lt srk (ArithTerm.eval srk termalg x) (ArithTerm.eval srk termalg y)
+      | open_formula -> Formula.construct srk open_formula
+    in
+    let reads_replaced = Formula.eval srk formalg body in
+    let functional_consistency_clauses =
+      List.map (fun (arr, read) ->
+          mk_if 
+            srk 
+            (mk_eq srk uq_term read)
+            (mk_eq srk (mk_const srk (uq_read arr)) (non_uq_read (arr, read))))
+        (BatHashtbl.to_list func_consist_reqs)
+    in
+    let matrix = mk_and srk (reads_replaced :: functional_consistency_clauses) in
+    let phi' = 
+      mk_exists_consts srk (fun sym -> not (Symbol.Set.mem sym !uqr_syms)) matrix 
+    in
+    let phi' = mk_forall_const srk uq_sym phi' in
+    mk_exists_consts srk (fun sym -> not (Symbol.Set.mem sym !nuqr_syms)) phi'
+
+
+  let pmfa_to_lia srk tf =
+    let tf = TransitionFormula.map_formula (eliminate_ite srk) tf in
+    let mfa, new_vars = to_mfa srk tf in
+    let lia = mfa_to_lia srk mfa in
+    let phi = 
+      mk_exists_consts srk (fun sym -> (not (Symbol.Set.mem sym new_vars))) lia
+    in
+    T.make ~exists:(T.exists tf) phi (T.symbols tf)
+
+  
+ let eliminate_stores srk phi =
+  let mk_op op =
+    match op with
+    | `Eq -> mk_eq
+    | `Lt -> mk_lt
+    | `Leq -> mk_leq
+  in
+  let rec rewrite_store index node =
+    match ArrTerm.destruct srk node with
+    | `Store (a, i, v) ->
+      let i = ArithTerm.eval srk arith_alg i in
+      let v = ArithTerm.eval srk arith_alg v in
+      mk_ite srk (mk_eq srk i index) v (rewrite_store index a)
+    | `Var (ind, `TyArr) -> mk_select srk (mk_var srk ind `TyArr) index
+    | `App (a, []) -> mk_select srk (mk_const srk a) index
+    | `Ite (phi, a, b) -> 
+      mk_ite 
+        srk 
+        (Formula.eval srk alg phi) 
+        (rewrite_store index a)
+        (rewrite_store index b)
+    | _ -> assert false
+  and  arith_alg = function
```

Comment:
You can use Syntax.rewrite (instead of Formula.eval / Term.eval) to make this more succinct.